src/session-pool.ts

```diff
@@ -542,62 +548,67 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
   }
 
   /**
-   * Attempts to create a session of a certain type.
+   * Attempts to create a single session of a certain type.
    *
    * @private
    *
    * @param {string} type The desired type to create.
    * @returns {Promise}
    */
-  async _createSession(type: types): Promise<void> {
-    const session = this.database.session();
-    const labels = this.options.labels!;
-
-    this._inventory.borrowed.add(session);
+  _createSession(type: types): Promise<void> {
+    const kind = type === types.ReadOnly ? 'reads' : 'writes';
+    const options = {[kind]: 1};
 
-    const createSession = async (): Promise<void> => {
-      await session.create({labels});
-
-      if (type === types.ReadWrite) {
-        try {
-          await this._prepareTransaction(session);
-        } catch (e) {
-          type = types.ReadOnly;
-        }
-      }
-    };
-
-    try {
-      await this._requests.add(createSession);
-    } catch (e) {
-      this._inventory.borrowed.delete(session);
-      throw e;
-    }
-
-    session.type = type;
-    session.lastUsed = Date.now();
-
-    this._inventory[type].push(session);
-    this._inventory.borrowed.delete(session);
+    return this._createSessions(options);
   }
 
   /**
-   * Attempts to create a session but emits any errors that occur.
+   * Batch creates sessions and prepares any necessary transactions.
    *
    * @private
    *
-   * @emits SessionPool#available
-   * @emits SessionPool#error
-   * @param {string} type The desired type to create.
+   * @param {object} [options] Config specifying how many sessions to create.
    * @returns {Promise}
    */
-  async _createSessionInBackground(type: types): Promise<void> {
+  async _createSessions({
+    reads = 0,
+    writes = 0,
+  }: CreateSessionsOptions): Promise<void> {
+    const count = reads + writes;
+    const labels = this.options.labels!;
+
+    this._pending += count;
+
+    let sessions: Session[] | null = null;
+
     try {
-      await this._createSession(type);
-      this.emit('available');
+      [sessions] = await this.database.batchCreateSessions({count, labels});
```

Comment:
It's possible that the number of sessions returned is less than what we ask for. So we need to do this in a while loop until the number of remaining sessions is zero.

Reply:
Thanks for the heads up, went ahead and made another commit to do just that.